internal/gps/pkgtree/pkgtree.go

```diff
@@ -921,3 +924,202 @@ func uniq(a []string) []string {
+	fi, er := os.Stat(vendorPathname)
+	if er != nil {
+		err = errors.Wrap(er, "cannot Stat")
+		return
+	}
+	if !fi.IsDir() {
+		err = errors.Errorf("cannot verify non directory: %q", vendorPathname)
+		return
+	}
+
+	status = make(map[string]LibraryStatus)
+	vendorPathname = filepath.Clean(vendorPathname) + pathSeparator
+	prefixLength := len(vendorPathname)
+
+	var otherNode *fsnode
+	currentNode := &fsnode{pathname: vendorPathname, parentIndex: -1, isRequiredAncestor: true}
+	nodes := []*fsnode{currentNode}
+	queue := []*fsnode{currentNode}
+
+	// NOTE: Mark directories of expected dependencies and their parents as required
+	for pathname := range expectedSums {
+		status[pathname] = NotInTree
+	}
+
+	for len(queue) > 0 {
+		// pop node from the queue (depth first traversal, reverse lexicographical order)
+		lq1 := len(queue) - 1
+		currentNode, queue = queue[lq1], queue[:lq1]
+
+		// log.Printf("NODE: %s", currentNode)
+		short := currentNode.pathname[prefixLength:] // chop off the prefix
+		if expectedSum, ok := expectedSums[short]; ok {
+			ls := EmptyDigestInLock
+			if expectedSum != "" {
+				ls = NoMismatch
+				actualSum, er := fs.HashFromNode(vendorPathname, short)
+				if err != nil {
+					err = errors.Wrap(er, "cannot compute dependency hash")
+					return
+				}
+				if actualSum != expectedSum {
+					ls = DigestMismatchInLock
+				}
+			}
+			status[short] = ls
+
+			// NOTE: Mark current nodes and all parents: required.
+			for pni := currentNode.myIndex; pni != -1; pni = otherNode.parentIndex {
+				otherNode = nodes[pni]
+				otherNode.isRequiredAncestor = true
+				// log.Printf("parent node: %s", otherNode)
+			}
+
+			continue // do not need to process directory's contents
+		}
+
+		childrenNames, er := sortedListOfDirectoryChildren(currentNode.pathname)
+		if er != nil {
+			err = errors.Wrap(er, "cannot get sorted list of directory children")
+			return
+		}
+		for _, childName := range childrenNames {
+			switch childName {
+			case ".", "..", "vendor", ".bzr", ".git", ".hg", ".svn":
+				// skip
+			default:
+				childPathname := filepath.Join(currentNode.pathname, childName)
+				otherNode = &fsnode{pathname: childPathname, myIndex: len(nodes), parentIndex: currentNode.myIndex}
+
+				fi, er := os.Stat(childPathname)
+				if er != nil {
+					err = errors.Wrap(er, "cannot Stat")
+					return
+				}
+				// Skip non-interesting file system nodes
+				mode := fi.Mode()
+				if mode&skipModes != 0 || mode&os.ModeSymlink != 0 {
+					// log.Printf("DEBUG: skipping: %v; %q", mode, currentNode.pathname)
+					continue
+				}
+
+				nodes = append(nodes, otherNode)
+				if fi.IsDir() {
+					queue = append(queue, otherNode)
+				}
+			}
+		}
+
+		if err != nil {
+			return // early termination iff error
+		}
+	}
+
+	// Walk nodes from end to beginning; ignore first node in list
+	for i := len(nodes) - 1; i > 0; i-- {
+		currentNode = nodes[i]
+		// log.Printf("FINAL: %s", currentNode)
```

Comment:
I will add a bit more embedded documentation to the answer *why* I create a tree of nodes, then walk them back to front.